feature/zimens/DP-31-SelectionMovies-detailsScreen

maxzim121 opened the pull request:
PR по выполненной задаче DP-31: разработка экрана подробностей о фильме.
• Свёрстан CustomMovieDetails
• Внесены изменения в экран выбора фильма
• Добавлена анимация по свайпу
• Изменена модель экрана выбора фильма

Close #31

maxzim121 committed:
Добавил необходимые файлы, начинаю работу над внедрением. Технический коммит

maxzim121 committed:
Merge branch 'develop' into feature/zimens/DP-31-SelectionMovies-detailsScreen

maxzim121 committed:
На экран выбора фильма добавлена view описания фильма по свайпу. Изменена модель данных для экрана.

maxzim121 committed:
Merge branch 'develop' into feature/zimens/DP-31-SelectionMovies-detailsScreen

maxzim121 committed:
Поправление кодстайла

maxzim121 committed:
Поправил кодстайл

IuriiKarpunov commented on DavaiPosmotrim/UI/CustomMovieDetails/CustomMovieDetails.swift:
```diff
@@ -0,0 +1,299 @@
+            ratingLabel.heightAnchor.constraint(equalToConstant: 24),
+
+            ratingCollection.topAnchor.constraint(equalTo: ratingLabel.bottomAnchor),
+            ratingCollection.leadingAnchor.constraint(equalTo: mainScroll.leadingAnchor, constant: 16),
```
коллекция должна при скрали уходить за экран, а если поставить отступ, то она будет обрезаться, поэтому 
ratingCollection.leadingAnchor.constraint(equalTo: mainScroll.leadingAnchor), и задать  
private lazy var ratingCollection: UICollectionView = {
...
layout.sectionInset = UIEdgeInsets(top: 0, left: 16, bottom: 0, right: 16)
...
}

maxzim121 replied:
✅

IuriiKarpunov commented on DavaiPosmotrim/UI/CustomMovieDetails/CustomRatingCollectionCell.swift:
```diff
@@ -0,0 +1,105 @@
+
+    private func configureBackground(number: Float) {
+        if number >= 7.0 {
+            contentView.backgroundColor = .lightSecondaryAccent
```
не тот цвет, нужен baseSecondaryAccent

maxzim121 replied:
✅

IuriiKarpunov commented on DavaiPosmotrim/UI/CustomMovieDetails/CustomRatingCollectionCell.swift:
```diff
@@ -0,0 +1,105 @@
+    
+    // MARK: - Public Methods
+
+    func configureKinopoiskRatingCell(ratingNumber: Float, votes: Int) {
+        rating.text = String(ratingNumber)
+        votesAmount.text = "\(String(votes)) оценок"
+        name.text = "Рейтинг Кинопоиска"
+        configureBackground(number: ratingNumber)
+    }
+
+    func configureIMDBRatingCell(ratingNumber: Float, votes: Int) {
+        rating.text = String(ratingNumber)
+        votesAmount.text = "\(String(votes)) оценок"
+        name.text = "Рейтинг IMDb"
+        configureBackground(number: ratingNumber)
+    }
```
string вынести в Resources

maxzim121 replied:
✅

IuriiKarpunov commented on DavaiPosmotrim/UI/CustomMovieDetails/CustomRatingCollectionCell.swift:
```diff
@@ -0,0 +1,105 @@
+
+    func configureKinopoiskRatingCell(ratingNumber: Float, votes: Int) {
+        rating.text = String(ratingNumber)
+        votesAmount.text = "\(String(votes)) оценок"
```
votes нужно разделять пробелом тысячные, можно использовать NumberFormatter()

maxzim121 replied:
✅

IuriiKarpunov commented on DavaiPosmotrim/UI/CustomMovieDetails/CustomMovieDetails.swift:
```diff
@@ -0,0 +1,299 @@
+    
+    // MARK: - Layout variables
+    
+    lazy var detailsText: UILabel = {
```
лучше сделать private, а взаимодействовать через методы

maxzim121 replied:
✅

IuriiKarpunov commented on DavaiPosmotrim/UI/CustomMovieDetails/CustomMovieDetails.swift:
```diff
@@ -0,0 +1,299 @@
+    private lazy var directorLabel: UILabel = {
+        var directorLabel = UILabel()
+        directorLabel.font = .textLabelFont
+        directorLabel.text = "Режиссёр"
```
вынести в ресурсы

maxzim121 replied:
✅

IuriiKarpunov commented on DavaiPosmotrim/UI/CustomMovieDetails/CustomMovieDetails.swift:
```diff
@@ -0,0 +1,299 @@
+            ratingCollection.leadingAnchor.constraint(equalTo: mainScroll.leadingAnchor, constant: 16),
+            ratingCollection.trailingAnchor.constraint(equalTo: trailingAnchor),
+            ratingCollection.heightAnchor.constraint(equalToConstant: 70)
+
+        ])
+
+    }
```
лишние отступы

maxzim121 replied:
✅

IuriiKarpunov commented on DavaiPosmotrim/UI/CustomMovieDetails/CustomMovieDetails.swift:
```diff
@@ -0,0 +1,299 @@
+    }
+
+    private func setupSubviews() {
+        [mainScroll, swipeView, slider].forEach {
```
если больше 2 параметров, тогда с новой строки

maxzim121 replied:
✅

IuriiKarpunov commented on DavaiPosmotrim/UI/CustomMovieDetails/CustomRatingCollectionCell.swift:
```diff
@@ -0,0 +1,105 @@
+        return name
+    }()
+    
+    //MARK: - Initializers
```
// MARK: - 
пробел забыл

maxzim121 replied:
✅

IuriiKarpunov reviewed (commented):
Хорошая работа, есть нюанс, когда закидываешь много текста, скрол работает некорректно, как будто высота высчитывается не правильно, попробуй, или открой на se и увидишь на втором фильме

superaydumov commented on DavaiPosmotrim/SelectionMoviesFlow/SelectionMovieDetailsCellModel.swift:
```diff
@@ -0,0 +1,18 @@
+import Foundation
+
+struct SelectionMovieDetailsCellModel {
+    let descreption: String
```
Опечатка в слове. Правильно будет "description"

maxzim121 replied:
✅

superaydumov commented on DavaiPosmotrim/SelectionMoviesFlow/SelectionMoviesViewController.swift:
```diff
@@ -138,10 +154,44 @@ private extension SelectionMoviesViewController {
+            return scene?.windows.first?.safeAreaInsets ?? .zero
+        }
+        let viewHeight = view.frame.height
+        let calculatedHeight = viewHeight - safeArea.top - customNavBar.frame.height - 16
```
Нужно ли тут учитывать ещё и safeArea.top? Если отступ у нас всё равно указан от навбара, который к этой safeArea и привязан?

superaydumov commented on DavaiPosmotrim/UI/CustomMovieDetails/CustomMovieDetails.swift:
```diff
@@ -0,0 +1,299 @@
+        return detailsText
+    }()
+    
+    lazy var swipeView: UIView = {
```
Здесь аналогично)) лучше сделать приватным, чтобы извне у нас были доступны только методы для настройки вью

maxzim121 replied:
✅

superaydumov commented on DavaiPosmotrim/UI/CustomMovieDetails/CustomMovieDetails.swift:
```diff
@@ -0,0 +1,299 @@
+        return mainScroll
+    }()
+
+    private lazy var slider: UIImageView = {
+        var slider = UIImageView(image: .slider)
+        return slider
+    }()
```
Можно укоротить использование вот так:

```
private lazy var slider = UIImageView(image: .slider)
```

maxzim121 replied:
✅

superaydumov commented on DavaiPosmotrim/UI/CustomMovieDetails/CustomMovieDetails.swift:
```diff
@@ -0,0 +1,299 @@
+
+    private lazy var detailsLabel: UILabel = {
+        var detailsLabel = UILabel()
+        detailsLabel.text = "Подробнее о фильме"
```
Строку лучше вынести в файл `Resources`

maxzim121 replied:
✅

superaydumov commented on DavaiPosmotrim/UI/CustomMovieDetails/CustomMovieDetails.swift:
```diff
@@ -0,0 +1,299 @@
+    private lazy var mainRolesLabel: UILabel = {
+        var mainRolesLabel = UILabel()
+        mainRolesLabel.font = .textLabelFont
+        mainRolesLabel.text = "В главных ролях"
```
Строку лучше вынести в файл `Resources`

maxzim121 replied:
✅

superaydumov commented on DavaiPosmotrim/UI/CustomMovieDetails/CustomMovieDetails.swift:
```diff
@@ -0,0 +1,299 @@
+        return directorCellBackground
+    }()
+
+    private lazy var directorCell: UILabel = {
```
Не совсем понятно...свойство называется directorCell, а тип имеет UILabel

P.S. Теперь понятно)) думаю, что нужно это переделать на ячейку коллекции. Потому что режиссёров может быть больше одного, но не более четырёх.
 
В этом случае у тебя будет две секции: одна с хэдером "В главных ролях", а другая - с хэдером "Режиссёр"

maxzim121 replied:
✅

superaydumov commented on DavaiPosmotrim/UI/CustomMovieDetails/CustomMovieDetails.swift:
```diff
@@ -0,0 +1,299 @@
+
+    private lazy var ratingLabel: UILabel = {
+        var ratingLabel = UILabel()
+        ratingLabel.text = "Оценки"
```
Строку лучше вынести в файл `Resources`

maxzim121 replied:
✅

superaydumov commented on DavaiPosmotrim/UI/CustomMovieDetails/CustomMovieDetails.swift:
```diff
@@ -0,0 +1,299 @@
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    // MARK: - Public Methods
+
+    override func layoutSubviews() {
+        super.layoutSubviews()
+        mainRolesCollectionView.layoutIfNeeded()
+        ratingCollection.layoutIfNeeded()
+    }
```
Этот метод же не относится в публичным, лучше перенести в Lifecycle

maxzim121 replied:
✅

superaydumov commented on DavaiPosmotrim/UI/CustomMovieDetails/CustomMovieDetails.swift:
```diff
@@ -0,0 +1,299 @@
+    func setupWithModel(model: SelectionMovieDetailsCellModel) {
+        detailsText.text = model.descreption
+        directorCell.text = model.directors[0]
+        mainRolesCollectionView.reloadData()
+        ratingCollection.reloadData()
```
Можешь использовать метод, который у тебя уже есть - `collectionsReloadData()`

maxzim121 replied:
✅

superaydumov commented on DavaiPosmotrim/UI/CustomMovieDetails/CustomMovieDetails.swift:
```diff
@@ -0,0 +1,299 @@
+
+import UIKit
+
+final class CustomMovieDetails: UIView, UICollectionViewDelegate {
```
А для чего используешь UICollectionViewDelegate?
Не увидел использования его методов

maxzim121 replied:
✅

maxzim121 replied:
✅

superaydumov commented on DavaiPosmotrim/UI/CustomMovieDetails/CustomRolesCollectionCell.swift:
```diff
@@ -0,0 +1,62 @@
+
+    // MARK: - Layout variables
+
+    lazy var roleLabel: UILabel = {
```
Свойство лучше сделать приватным.
У тебя же есть метод для его настройки

maxzim121 replied:
✅

superaydumov commented on DavaiPosmotrim/UI/CustomMovieDetails/CustomRolesCollectionCell.swift:
```diff
@@ -0,0 +1,62 @@
+
+    override init(frame: CGRect) {
+        super.init(frame: frame)
+        contentView.layer.cornerRadius = 12
+        contentView.backgroundColor = .baseBackground
```
В качестве рекомендации:
На мой взгляд можно объединить методы setupSubviews и setupConstraints, и перенести эти строки в новый метод.

Назвать его можно что-то вроде setupCell. В ячейке по сути просто один лэйбл и можно в данном случае всё объединить

maxzim121 replied:
✅

superaydumov reviewed (commented):
Максим, оставил свои комментарии и предложения по улучшению!))
По работе на SE..запустил у себя и информация не влезает, да. Высоты скроллвью недостаточно

superaydumov commented:
Кстати, в качестве предложения по поводу коллекций.

Можешь использовать переиспользуемую коллекцию в папке UI. Только прикинуть в неё свою ячейку и всё должно заработать..поможет немного сократить код.

maxzim121 committed:
Частично исправил замечания по ПР

maxzim121 committed:
Объеденил коллекцию Главных ролей и Режиссёров в одну. Добавил файл LeftAllignedLayout для правильного отображения ячеек

maxzim121 committed:
Переделал реализацию CustomMovieDetails. Избавился от лишних эллементов. Исправил баг отображения на SE. Остался баг отображения рейтингов на маленьких экранах при скролле.
Добавил файл LeftAllignLayout для выравнивания ячеек по левой стороне c UICollectionViewFlowLayout.
Также исправлен баг SplashScreen для 15 Pro Max.

maxzim121 committed:
Исправил баг отобржаения коллекции.

superaydumov commented on DavaiPosmotrim/UI/CustomMovieDetails/CustomMovieDescriptionCell.swift:
```diff
@@ -0,0 +1,55 @@
+
+    // MARK: - Layout variables
+
+    private lazy var discriptionLabel: UILabel = {
```
Правильно название - descriptionLabel

maxzim121 replied:
✅

superaydumov commented on DavaiPosmotrim/UI/CustomMovieDetails/CustomMovieDetails.swift:
```diff
@@ -0,0 +1,279 @@
+        return actorsAndDirectirsCollectionView
+    }()
+
+    private lazy var ratingLabel: UILabel = {
+        var ratingLabel = UILabel()
+        ratingLabel.text = Resources.MovieDetails.votesText
+        ratingLabel.font = .textLabelFont
+        return ratingLabel
+    }()
```
Это вычисляемое свойство нигде не используется

maxzim121 replied:
✅

superaydumov commented on DavaiPosmotrim/UI/CustomMovieDetails/CustomMovieDetails.swift:
```diff
@@ -0,0 +1,279 @@
+
+    // MARK: - Public Methods
+
+    func collectionsReloadData() {
+        actorsAndDirectirsCollectionView.reloadData()
+    }
```
Лучше вместо этого метода использовать просто:
actorsAndDirectorsCollectionView.reloadData()

maxzim121 replied:
✅

maxzim121 replied:
В одном моменте этот метод нужен, при первом свайпе обновляю коллекцию чтобы она правильно отображалась

superaydumov commented on DavaiPosmotrim/UI/CustomMovieDetails/CustomMovieDetails.swift:
```diff
@@ -0,0 +1,279 @@
+        collectionsReloadData()
+    }
+
+    func setupWithModel(model: SelectionMovieDetailsCellModel) {
+        collectionsReloadData()
+    }
```
Не очень понимаю...для чего нужен этот метод?

maxzim121 replied:
✅

superaydumov commented on DavaiPosmotrim/UI/CustomMovieDetails/CustomMovieDetails.swift:
```diff
@@ -0,0 +1,279 @@
+
+    private lazy var slider: UIImageView = UIImageView(image: .slider)
+
+    private lazy var actorsAndDirectirsCollectionView: UICollectionView = {
```
Правильное название - actorsAndDirectorsCollectionView

maxzim121 replied:
✅

superaydumov commented on DavaiPosmotrim/UI/CustomMovieDetails/CustomMovieDescriptionCell.swift:
```diff
@@ -0,0 +1,55 @@
+            discriptionLabel.leadingAnchor.constraint(equalTo: contentView.leadingAnchor),
+            discriptionLabel.topAnchor.constraint(equalTo: contentView.topAnchor),
+            discriptionLabel.bottomAnchor.constraint(equalTo: contentView.bottomAnchor),
+            discriptionLabel.widthAnchor.constraint(equalToConstant: contentView.frame.width - 32)
```
Можно просто задать trailingAnchor c отступом, чтобы не высчитывать ширину

maxzim121 replied:
✅

superaydumov reviewed (commented):
Максим, оставил несколько комментов внутри!)) хорошее решение проблемы с коллекцией

maxzim121 committed:
Исправил замечания по ПР.

maxzim121 committed:
Исправил отображение коллекции рейтингов после свайпа

IuriiKarpunov reviewed (commented):
ты перепутал местами "Подробнее о фильме" и "В главных ролях")

maxzim121 committed:
Добавил метод для прокручивания коллекции вверх после того, как пользователь сворачивает экран подробностей. Исправил хедеры

uistelmakh commented on DavaiPosmotrim/SelectionMoviesFlow/SelectionMoviesViewController.swift:
```diff
@@ -138,10 +153,45 @@ private extension SelectionMoviesViewController {
 
+    func changeDirection(direction: UISwipeGestureRecognizer.Direction) -> UISwipeGestureRecognizer.Direction {
+        if direction == .up {
+            self.customMovieDetails.collectionReloadData()
```
self не нужен

maxzim121 replied:
✅

uistelmakh commented on DavaiPosmotrim/SelectionMoviesFlow/SelectionMoviesViewController.swift:
```diff
@@ -138,10 +153,45 @@ private extension SelectionMoviesViewController {
+            UIView.animate(withDuration: 0.3, animations: {
+                self.customMovieDetails.transform = CGAffineTransform(translationX: 0, y: 0)
+            })
+            self.customMovieDetails.collectionReloadData()
```
здесь тоже

maxzim121 replied:
✅

uistelmakh commented on DavaiPosmotrim/UI/CustomMovieDetails/CustomMovieDetailsCollectionHeader.swift:
```diff
@@ -0,0 +1,41 @@
+    func configure(text: String) {
+        label.text = text
+    }
+
```
отступ

maxzim121 replied:
✅

maxzim121 committed:
Исправил комментарии по ПР
